shiny/types.py

```diff
@@ -31,24 +31,30 @@
 from htmltools import TagChild
 
 from ._docstring import add_example
-from ._typing_extensions import NotRequired, TypedDict
+from ._typing_extensions import NotRequired, TypedDict, TypeIs
 
 if TYPE_CHECKING:
     from matplotlib.figure import Figure
 
+T = TypeVar("T")
+
 
 # Sentinel value - indicates a missing value in a function call.
 class MISSING_TYPE:
     pass
 
 
 MISSING: MISSING_TYPE = MISSING_TYPE()
+DEPRECATED: MISSING_TYPE = MISSING_TYPE()  # A MISSING that communicates deprecation
+MaybeMissing = Union[T, MISSING_TYPE]
 
-
-T = TypeVar("T")
 ListOrTuple = Union[List[T], Tuple[T, ...]]
 
 
+def is_missing(x: Any) -> TypeIs[MISSING_TYPE]:
+    return x is MISSING or isinstance(x, MISSING_TYPE)
```

Comment:
TIL about the difference between `TypeGuard` and `TypeIs`. (`TypeGuard` only narrows the type scope in the **positive** case [if it is, it is, otherwise it might still be], while `TypeIs` narrows the type on both sides [it either is or it isn't].)

Anyway, this could also just be `isinstance(x, MISSING_TYPE)` but it's a little more succinct when used.